pkg/operator/ceph/file/mds/mds.go

```diff
@@ -120,86 +133,168 @@ func (c *Cluster) Start() error {
+		}
+		if cephver.IsSuperior(c.clusterInfo.CephVersion, *currentVersion) {
+			logger.Debugf("ceph version for MDS %q is %q and mon version is %q", key, currentVersion, c.clusterInfo.CephVersion)
+			return true, err
 		}
-		desiredDeployments[d.GetName()] = true // add deployment name to improvised set
+	}
+
+	return false, nil
+}
+
+func (c *Cluster) upgradeMDS() error {
 
+	// 1. set allow_standby_replay to false
+	if err := cephclient.AllowStandbyReplay(c.context, c.clusterInfo, c.fs.Name, false); err != nil {
+		return errors.Wrapf(err, "failed to setting allow_standby_replay to false")
 	}
 
-	if err := c.scaleDownDeployments(replicas, desiredDeployments); err != nil {
-		return errors.Wrap(err, "failed to scale down mds deployments")
+	// In Pacific, standby-replay daemons are stopped automatically. Older versions of Ceph require us to stop these daemons manually.
+	if err := cephclient.FailAllStandbyReplayMDS(c.context, c.clusterInfo, c.fs.Name); err != nil {
+		return errors.Wrapf(err, "failed to fail mds agent in up:standby-replay state")
+	}
+
+	// 2. set max_mds to 1
+	logger.Debug("start setting active mds count to 1")
+	if err := cephclient.SetNumMDSRanks(c.context, c.clusterInfo, c.fs.Name, 1); err != nil {
+		logger.Warningf("failed setting active mds count to %d. %v", 1, err)
+		return err
+	}
+	// 3. wait for ranks to be 0
+	if err := cephclient.WaitForActiveRanks(c.context, c.clusterInfo, c.fs.Name, 1, false, fsWaitForActiveTimeout); err != nil {
+		return errors.Wrapf(err, "failed waiting for active ranks to be 1")
+	}
+	// 4. stop standby nodes
+	daemonName, err := cephclient.GetMdsIdByRank(c.context, c.clusterInfo, c.fs.Name, 0)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get mds id from rank 0")
+	}
+	daemonNameTokens := strings.Split(daemonName, "-")
+	daemonLetterID := daemonNameTokens[len(daemonNameTokens)-1]
+	desiredDeployments := map[string]bool{
+		fmt.Sprintf("%s-%s-%s", AppName, c.fs.Name, daemonLetterID): true,
 	}
+	logger.Debugf("stop mds other than %s", daemonName)
+	err = c.scaleDownDeployments(1, 1, desiredDeployments, false)
+	if err != nil {
+		return errors.Wrapf(err, "failed to scale down deployments during upgrade")
+	}
+	// 5. upgrade current active deployment and wait for it come back
+	_, err = c.startDeployment(context.TODO(), daemonLetterID)
+	if err != nil {
+		return errors.Wrapf(err, "failed to upgrade mds %s", daemonName)
+	}
+	logger.Debugf("successfully started daemon %s", daemonName)
 
+	// 6. set max_mds back to what it should be for user config
+	if err := cephclient.AllowStandbyReplay(c.context, c.clusterInfo, c.fs.Name, true); err != nil {
+		return errors.Wrapf(err, "failed to setting allow_standby_replay to true")
+	}
+	if err := cephclient.SetNumMDSRanks(c.context, c.clusterInfo, c.fs.Name, c.fs.Spec.MetadataServer.ActiveCount); err != nil {
+		return errors.Wrapf(err, "failed to set max_mds to %d", c.fs.Spec.MetadataServer.ActiveCount)
+	}
 	return nil
```

Comment:
I think a comment to clarify when other daemons are updated is good here.

```suggestion
	}
	
	// 7. all other MDS daemons will be updated and restarted by main MDS code path
	
	return nil
```